### ISSUE TITLE
[HOLD - #20558][$1000] Anonymous User Access and Account Claiming in Public Chat Rooms

### ISSUE THREAD

**Issue by marcochavezf** [Expensify/App#18321]
We'll introduce a type of user called `anonymous account`. This account will be identified by a new property `authTokenType` added to the object `session`:

<img width="556" alt="Screenshot 2023-05-10 at 16 41 16" src="https://github.com/Expensify/App/assets/6829422/acb98073-bbd4-4e15-9653-a6333ed0a2e4">

This user will see the application in read-only mode (they can't post or trigger any action that could persist data in the server). 

![Chat - Desktop (3)](https://user-images.githubusercontent.com/6829422/235811137-f0ca9216-cded-4b41-8c74-c64e154e1149.png)

These users can use the application like any other user, but they will encounter a sign-in page when attempting to perform certain actions. We are currently building functionality for this account type and will prevent anonymous users from executing specific server commands. In the meantime, we will implement changes to the front-end in preparation for anonymous accounts.

We'll create a new function called isAnonymousUser, which will check if `session.authTokenType === 'anonymousAccount'`. This function will be used to implement the following features (to simulate an anonymous account, we can always return true until `session.authTokenType` is ready):

1. A sign-in box will appear at the bottom of the page (replacing the compose box) to indicate that the user is accessing the room with an anonymous account (this will also apply to the Concierge chat):
2. The `Sign In` button will open the sign-in page as full-screen modal dialog with an X button to close the sign-in page and show the public room again.
3. The search functionality will work as normal for the user.
4. We’ll allow the user to open the Global Create (“+”) button, but once the user clicks on one of the options, the sign-in page will be shown.
5. The “Start a call” and “Pin” buttons will also open the sign-in page.
6. The anonymous user will be able to open the details of the public room, once they click on one of the options (“Members” or “Settings”) the sign-in page will be shown.
7. We will also show the sign-in page if the user wants to add a reaction or trigger any action on a message.
8. In the new sign-in page when the user press the "Continue" button, we'll call new action named 'claimAccount' (instead of [beginSignIn](https://github.com/Expensify/App/blob/main/src/libs/actions/Session/index.js#L207-L256)). It will call the command `ClaimAnonymousAccount`.

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01ccde2cd289cc869f</li>
        <li>Upwork Job ID: 1653552921324421120</li>
        <li>Last Price Increase: 2023-05-12</li>
    </ul>
</details>

**Comment by MelvinBot** [comment]
Job added to Upwork: https://www.upwork.com/jobs/~01ccde2cd289cc869f

**Comment by MelvinBot** [comment]
Triggered auto assignment to Contributor Plus for review of internal employee PR - @sobitneupane (`Internal`)

**Comment by marcochavezf** [comment]
Still on hold

**Comment by marcochavezf** [comment]
Making this GH external in case someone wants to propose a solution while we complete the backend implementation.

**Comment by alexpensify** [comment]
Sounds like a plan @marcochavezf

**Comment by marcochavezf** [comment]
Not overdue, waiting for proposals

**Comment by Swor71** [comment]
Hey, I'm Marcin from Callstack - expert contributor group - I would like to work on this issue.

**Comment by thiagobrez** [comment]
And I'll be helping @Swor71 ⬆️

**Comment by Swor71** [comment]
Daily update: I'm working with Marco to resolve a few questions I have around this issue and doing some investigation in the meantime.

**Comment by alexpensify** [comment]
@dylanexpensify - this one is moving forward, but assigning if a BZ team member is needed to step in while I'm OOO.

**Comment by dylanexpensify** [comment]
@Swor71 can we get an update please? 🙇‍♂️

**Comment by Swor71** [comment]
@dylanexpensify daily update: I've done the modal part, reused the sign in page, at the moment I'm implementing it in the places mentioned in the list above

**Comment by dylanexpensify** [comment]
[I'm heading OOO tomorrow](https://expensify.slack.com/archives/C01SKUP7QR0/p1684776773511559) for about a week so adding another BZ member while I'm away to help keep the train moving! 🚂 Thanks @miljakljajic!!

**Comment by thiagobrez** [comment]
Hey! I'm Thiago from Callstack - expert contributor group - and I'll be taking this over from @Swor71 as he will be OOO for medical reasons. He already briefed me with everything needed and it's looking good. Will come with more updates later today :)

**Comment by thiagobrez** [comment]
Update: Bullet points 2-6 are done. Working on 7. Then, 8 should be simple enough. Leaving 1 to the end.

**Comment by thiagobrez** [comment]
Update: Still working on covering all actions on messages. There's quite a few situations, both on web and mobile. Unfortunately, I'll be out of office this Thursday and Friday. Will continue on Monday.

**Comment by alexpensify** [comment]
@miljakljajic and @dylanexpensify - I appreciate your help here! I'm back online and removing your assignment.

**Comment by alexpensify** [comment]
Thank you for the update @thiagobrez for the update!

**Comment by marcochavezf** [comment]
Hi @thiagobrez @Swor71, is it possible for you to push all your changes into a WIP/draft PR?

**Comment by thiagobrez** [comment]
Hey @marcochavezf! I'm away from the computer right now, can do it tomorrow morning 👍🏻

**Comment by thiagobrez** [comment]
@marcochavezf Draft PR is here: https://github.com/Expensify/App/pull/19598 🙌🏻 

A few notes:

>1. A sign-in box will appear at the bottom of the page (replacing the compose box) to indicate that the user is accessing the room with an anonymous account (this will also apply to the Concierge chat):

Not done yet

>2. The Sign In button will open the sign-in page as full-screen modal dialog with an X button to close the sign-in page and show the public room again.

Done, but since there were no design references, need to check if it is up to your liking

>3. The search functionality will work as normal for the user.

N/A

>4. We’ll allow the user to open the Global Create (“+”) button, but once the user clicks on one of the options, the sign-in page will be shown.

Done

>5. The “Start a call” and “Pin” buttons will also open the sign-in page.

Done

>6. The anonymous user will be able to open the details of the public room, once they click on one of the options (“Members” or “Settings”) the sign-in page will be shown.

Done

>7. We will also show the sign-in page if the user wants to add a reaction or trigger any action on a message.

Done. Only one caveat here is that I noticed that when triggering reactions, the reaction is not sent (OK) but another API call to `UpdateFrequentlyUsedEmojis` is being sent. I was working on this but did not finish.

>8. In the new sign-in page when the user press the "Continue" button, we'll call new action named 'claimAccount' (instead of [beginSignIn](https://github.com/Expensify/App/blob/main/src/libs/actions/Session/index.js#L207-L256)). It will call the command ClaimAnonymousAccount.

Not done yet

Let me know if you need anything else!

**Comment by marcochavezf** [comment]
Thank you @thiagobrez for the PR and update! Given the urgency, let's keep using `beginSignIn` in step 8 
The implementation of claimAccount is in review, but it will probably take some to be deployed, so let's just re-use the sign-in page and let the user create a new account as it happens now. We will change it to claimAccount as a follow-up PR when the backend changes are deployed (which I estimate will happen next week).

**Comment by alexpensify** [comment]
This one is under review!

**Comment by marcochavezf** [comment]
Hey @thiagobrez! We're switching up the requirements for this issue. I've incorporated some of your changes (big thanks for that!) into this [PR](https://github.com/Expensify/App/pull/18293) and tweaked them to meet a deadline. We'll now stick to the original plan of showing the sign-in page in a modal and using the ClaimAnonymousAccount command (which should go live sometime this week). On top of that, I'll drop the requirement to implement the sign-in box.

We've also changed how we check if an anonymous user is allowed to perform an action in the linked PR. So, you might want to adapt the changes here to align with this new approach. We're now verifying in Session/index.js if an action is allowed by an anonymous user, instead of subscribing to the Onyx session key in different components to avoid unnecessary component re-rendering.

**Comment by thiagobrez** [comment]
Hey @marcochavezf! My pleasure, glad you guys got it working. Just checked the PR and looks good 🙌🏻 

>We'll now stick to the original plan of showing the sign-in page in a modal and using the ClaimAnonymousAccount command (which should go live sometime this week). On top of that, I'll drop the requirement to implement the sign-in box.

Ok, so I'll pull those latest changes and continue here to work on the above ⬆️

**Comment by thiagobrez** [comment]
Is the Sign-in Page any different when in the "Claim account" flow? Or still the same: email -> magic code?

**Comment by marcochavezf** [comment]
Hi @thiagobrez thanks! 

> Is the Sign-in Page any different when in the "Claim account" flow? Or still the same: email -> magic code?

Good question, that's something that's not completely defined. I will copy-paste what I replied via DM and let me know your thoughts!

> What’s still in the air is what will happen in the UI after the user triggers the ClaimAnonymousAccount command. The last thing that we agreed on was to remove the “X” button from the modal after the user enters their email and then show the validate the code screen, but tbh I’m not sure if that will work.
>
> I’m not sure how things will work since we’ll still be displaying the public chat and LHN (left hand navigator) in the background. It might feel a bit odd to keep showing the validation code page, and then we’d need to refresh both the public chat and the LHN after the user enters their validation code
>
> On the other hand, since the anonymous user will now be the authenticated user, we can keep the LHN and public rooms as is and only change the authToken and email.
>
> Another idea is that after the user enters their email, we sign out the anonymous user, close the modal and redirect the user to the validate code page.
> 
>But I think we would need to prototype something once the ClaimAnonymousAccount command is completed. Meanwhile we can work to display the sign-in page in the modal (that will also help as part of the prototype to figure out what will be the best route once the new command is ready)
> 
> Let me know any doubts or if I’m missing something :slightly_smiling_face:

**Comment by thiagobrez** [comment]
>The last thing that we agreed on was to remove the “X” button from the modal after the user enters their email and then show the validate the code screen, but tbh I’m not sure if that will work.

Yeah that seems like an easy way to bypass / interrupt the flow, IMO

>I’m not sure how things will work since as we’ll still be displaying the public chat and LHN (left hand navigator) in the background. It might feel a bit odd to keep showing the validation code page, and then we’d need to refresh both the public chat and the LHN after the user enters their validate code

That's my preferred option. I think that if I'm logging in inside a Modal, most natural thing would be to finish the flow inside the modal. Then, refreshing the data that was in background seems like a good idea.

I have a [new draft PR here](https://github.com/Expensify/App/pull/19808) with the Modal implementation, and I'll try these scenarios. But I'm having trouble testing because I'm currently not able to open a public room locally, and it's not quite the same as [Vit reported here](https://github.com/Expensify/App/pull/18293#issuecomment-1568392572).

For example, if I sign out, then get the ExpensiConX url (replaced with localhost) and paste in, the app will just reload in the sign in page. If I then try to log in, that's when I get the error Vit reported. Am I missing anything?

**Comment by marcochavezf** [comment]
Hello @thiagobrez! We've pushing new changes lately and some of the errors have been fixed. Alternatively, you could use also this public room for testing https://staging.new.expensify.com/r/6469015542180863

**Comment by alexpensify** [comment]
No update, this one looks to be moving forward.

**Comment by thiagobrez** [comment]
Sorry for the lack of update, I'm currently occupied with other issues. Will need to revisit this on Monday.

**Comment by alexpensify** [comment]
All good, thank you for the update @thiagobrez

**Comment by alexpensify** [comment]
Based on the last update, there will be more news early this week.

**Comment by thiagobrez** [comment]
I got a working solution in the draft PR and it's being reviewed internally.

Last thing would be to change the sign in command to the `ClaimAnonymousAccount` command. Do you happen to know the status of it @alexpensify ?

**Comment by marcochavezf** [comment]
Hi @thiagobrez! We're still working on the `ClaimAnonymousAccount`, I hope it will be ready this week. I let you know once it's ready. Meanwhile, I will use the draft PR to test all the pieces together and let you know if we need to change something else. So we're good atm, thanks!

**Comment by thiagobrez** [comment]
Awesome @marcochavezf! No rush, I don't think that's a blocker. I feel like it's already working well with the SignIn command, but let me know what you think. Thanks!

**Comment by thiagobrez** [comment]
Made the PR ready for review here:

https://github.com/Expensify/App/pull/19808

**Comment by marcochavezf** [comment]
Hi @shawnborton! When you're back, could you take a look at the design of the modal in this [PR](https://github.com/Expensify/App/pull/19808) (videos)?

**Comment by alexpensify** [comment]
No update, we are waiting on design feedback here.

**Comment by shawnborton** [comment]
Just noting that I am addressing the design feedback in the linked PR above.

**Comment by alexpensify** [comment]
🤦🏼 - sorry Shawn! I didn't check the PR.

**Comment by alexpensify** [comment]
We are moving forward here!

**Comment by alexpensify** [comment]
The PR is still being reviewed.

**Comment by alexpensify** [comment]
The PR is being worked on.

**Comment by alexpensify** [comment]
Still in review

**Comment by marcochavezf** [comment]
Update: Active discussion about the solution [here](https://expensify.slack.com/archives/C01GTK53T8Q/p1686787152256199)

**Comment by alexpensify** [comment]
Thank you for that update. I'll keep watching for a summary in that Slack convo.

**Comment by alexpensify** [comment]
@marcochavezf can we get an update here? Thank you!

**Comment by thiagobrez** [comment]
@alexpensify @marcochavezf Hey! Seems like the approach to joining navigation stacks is going forward. Should I close [my old PR with the modal implementation](https://github.com/Expensify/App/pull/19808)?

**Comment by daordonez11** [comment]
Hey @alexpensify  just left an update comment in https://github.com/Expensify/App/issues/20558, I am thinking we can split the problem in 2 things:
1. Navigation flow, how to handle login if a user is in a public room.
2. Anon user implementations such as claimaccount 

I think a lot of the modal code is not required for the signinpage in the RHP but definitely the props and the implementation @thiagobrez did for claim is needed. Maybe @thiagobrez can upload the implementation of supporting `isAnonymous` in the signInPage and I can focus on the navigation side.

What do you guys think?

cc @marcochavezf

**Comment by marcochavezf** [comment]
Yeah, the work @thiagobrez has done on the other PR has been super helpful for the implementation of anonymous users. I agree that the props setup will still be necessary. I still have to make some backend changes for the account claim feature, and I've been waiting until we've decided on a clear path for the solution. This choice might affect whether we'll need to clear the anonymous user's session from the backend when the new account claim command is called. With that in mind, I'll put on hold the other PR (which will be part of the solution for point 2 that @daordonez11 stated above) until the RHP implementation and account claim command are completed.

**Comment by alexpensify** [comment]
@marcochavezf since this is going on hold, should we make it a weekly?

**Comment by marcochavezf** [comment]
Yeah, that makes sense. Changing it to weekly

**Comment by alexpensify** [comment]
Reassigning another `bug` team member, I'm going OOO until Monday, July 10, and will take it back then if it's not closed by then. 

Required action from the 🐛 team: Check to see if this one goes back to a daily and we can resume work here.

**Comment by alexpensify** [comment]
Alright, removing the other 🐛 team member. I'm back online and see that this is our current state: 

Still on hold

**Comment by alexpensify** [comment]
On hold

**Comment by alexpensify** [comment]
Weekly Update: On hold

**Comment by alexpensify** [comment]
@marcochavezf - can you share the link to the GH that we are tracking to unlock this one? I realized that I haven't added it to the title. Thank you!

**Comment by daordonez11** [comment]
@alexpensify it would be https://github.com/Expensify/App/issues/20558

**Comment by alexpensify** [comment]
Thank you for that update @daordonez11!

**Comment by alexpensify** [comment]
Weekly Update: On hold

**Comment by alexpensify** [comment]
Weekly update: On hold

**Comment by alexpensify** [comment]
@marcochavezf and @sobitneupane - it looks like https://github.com/Expensify/App/issues/20558 is in staging. Should we remove the hold on this one and start accepting proposals again? Thanks!

**Comment by daordonez11** [comment]
@alexpensify actually I think this was already solved by @marcochavezf in backend finally it was solved in the Signin command so I think we could close both issues only with the sign in in RHP.

**Comment by alexpensify** [comment]
Nice! @marcochavezf can you confirm and close? Thanks!

**Comment by alexpensify** [comment]
I think @marcochavezf is OOO, so going to close for now.
